### PR TITLE
Parse package specifier before install

### LINKED ISSUE #438: Installing VCS URL should not use `--editable` option
**Describe the bug**

Our package name detection scheme depends on installing a package first with pip and querying its name.  pip seems to allow installation with a bare VCS url **ONLY** if the `--editable` flag is not present.  If the `--editable` flag is present, then pip seems to require the suffix `#egg=<package_name>`.

**How to reproduce**

Using a git package from another issue I was reading (#148) and trying to reproduce:

```
birchtree:pipx pipx install --verbose --editable git+https://github.com/raxod502/diary-manager
pipx > (run_subprocess:112): running /usr/local/opt/python@3.8/bin/python3.8 -m venv --without-pip /var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli
pipx > (run_subprocess:112): running /var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli/bin/python -c import sysconfig; print(sysconfig.get_path('purelib'))
pipx > (run_subprocess:112): running /Users/mclapp/.local/pipx/shared/bin/python -c import sysconfig; print(sysconfig.get_path('purelib'))
pipx > (run_subprocess:112): running /var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli/bin/python --version
pipx > (run_subprocess:112): running /var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli/bin/python -m pip list --format=json
pipx > (run_subprocess:112): running /var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli/bin/python -m pip install --no-dependencies --editable git+https://github.com/raxod502/diary-manager
ERROR: Could not detect requirement name for 'git+https://github.com/raxod502/diary-manager', please specify one with #egg=your_package_name
pipx > (install_package_no_deps:221): '/var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/tmp_5cntfli/bin/python -m pip install --no-dependencies --editable git+https://github.com/raxod502/diary-manager' failed
Cannot determine package name from spec 'git+https://github.com/raxod502/diary-manager'. Check package spec for errors.
```

Just with pip itself, `--editable` + VCS URL needs `#egg`:
```
birchtree:pipx pip install --no-dependencies --editable git+https://github.com/raxod502/diary-manager
ERROR: Could not detect requirement name for 'git+https://github.com/raxod502/diary-manager', please specify one with #egg=your_package_name
```
Just with pip itself, bare VCS URL does NOT need `#egg`:
```
birchtree:pipx pip install --no-dependencies git+https://github.com/raxod502/diary-manager 
Collecting git+https://github.com/raxod502/diary-manager
  Cloning https://github.com/raxod502/diary-manager to /private/var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/pip-req-build-5kt0mjb7
  Running command git clone -q https://github.com/raxod502/diary-manager /private/var/folders/gp/7z0mcw717y1b0kt0swc1lfp80000gn/T/pip-req-build-5kt0mjb7
Using legacy setup.py install for diary-manager, since package 'wheel' is not installed.
Installing collected packages: diary-manager
    Running setup.py install for diary-manager ... done
Successfully installed diary-manager-2.0.2
```

**Expected behavior**

I expected pip could install the VCS URL whether it was --editable or not.

### PR DESCRIPTION
**Pull request by itsayellow**
[x] I have added an entry to `docs/changelog.md`

## Summary of changes
Fixes #438 .

This parses a package specifier before install.

It removes the `--editable` option for any package specification that is not a local path, and gives a warning that it is doing so.
It removes any environment markers from the package specification and gives a warning that it is doing so.
I also refactored `package_specifier.py` a bit.
And a bunch of tests were added for the new functionality.

I started this PR as a fix for a pip error within pipx that was complicated enough that there might be user confusion.  (See #438).  Once I was parsing the package specifier, I also decided to remove the environment markers that cause errors in pipx operation, and anyway seemed to me not a fit for pipx's use-case.

### Test 1
```
pipx install -e pylint
```
This PR gives a warning that pipx is removing the `--editable` option.  The current dev version of pipx would just pass-through a pip error.

### Test 2
```
pipx install --editable git+https://github.com/raxod502/diary-manager
```
This PR gives a warning that pipx is removing the `--editable` option.  The current dev version of pipx would just pass-through an even more complicated pip error.

### Test 3
```
pipx install 'pylint;python_version=="3.6"'
```
Current dev pipx gives cryptic error:
```
Internal error with venv metadata inspection.

Unable to install pylint from spec 'pylint;python_version=="3.6"'.
Check the name or spec for errors, and verify that it can be installed with pip.
```
This PR pipx gives warning it is ignoring markers and proceeds normally (without error!) to install pylint.

### CODE REVIEW

**Comment by itsayellow** [comment]
I just added another parse inside of `commands/common.py` in `package_name_from_spec`, another place besides `venv.py` that uses a package spec. that I previously missed.

This pre-parses the package spec to check syntax before we send it to pip, and to also clean the spec.

Here is what happens if you give a bad syntax package spec to the PR with my latest commit:
```
>:pipx pipx install 'pylint;python_version="3.6"'
Unable to parse package spec: pylint;python_version="3.6"
>:pipx
```
Here is what happens if you give a bad syntax package spec to the current pipx.  A page of stack trace from pip, all in red.
```
>: pipx install 'pylint;python_version="3.6"' 
ERROR: Exception:kage name from 'pylint;python_version="3.6"'
Traceback (most recent call last):
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_vendor/packaging/markers.py", line 299, in __init__
    self._markers = _coerce_parse_result(MARKER.parseString(marker))
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_vendor/pyparsing.py", line 1955, in parseString
    raise exc
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_vendor/pyparsing.py", line 2893, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pip._vendor.pyparsing.ParseException: Expected {Group:({{{"implementation_version" | "platform_python_implementation" | "implementation_name" | "python_full_version" | "platform_release" | "platform_version" | "platform_machine" | "platform_system" | "python_version" | "sys_platform" | "os_name" | "os.name" | "sys.platform" | "platform.version" | "platform.machine" | "platform.python_implementation" | "python_implementation" | "extra"} | {quoted string, starting with ' ending with ' | quoted string, starting with " ending with "}} {"===" | "==" | ">=" | "<=" | "!=" | "~=" | ">" | "<" | "not in" | "in"} {{"implementation_version" | "platform_python_implementation" | "implementation_name" | "python_full_version" | "platform_release" | "platform_version" | "platform_machine" | "platform_system" | "python_version" | "sys_platform" | "os_name" | "os.name" | "sys.platform" | "platform.version" | "platform.machine" | "platform.python_implementation" | "python_implementation" | "extra"} | {quoted string, starting with ' ending with ' | quoted string, starting with " ending with "}}}) | Group:({{Suppress:("(") : ...} Suppress:(")")})}, found '='  (at char 14), (line:1, col:15)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 188, in _main
    status = self.run(options, args)
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/cli/req_command.py", line 185, in wrapper
    return func(self, options, args)
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/commands/install.py", line 300, in run
    reqs = self.get_requirements(
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/cli/req_command.py", line 321, in get_requirements
    req_to_add = install_req_from_line(
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/req/constructors.py", line 396, in install_req_from_line
    parts = parse_req_from_line(name, line_source)
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_internal/req/constructors.py", line 312, in parse_req_from_line
    markers = Marker(markers_as_string)
  File "/Users/mclapp/.local/pipx/shared/lib/python3.8/site-packages/pip/_vendor/packaging/markers.py", line 304, in __init__
    raise InvalidMarker(err_str)
pip._vendor.packaging.markers.InvalidMarker: Invalid marker: 'python_version="3.6"', parse error at '="3.6"'
Cannot determine package name from spec 'pylint;python_version="3.6"'. Check package spec for errors.
```

**Comment by itsayellow** [comment]
Additionally I refactored so all package name determination (even in `run`) now goes through `commands.common.package_name_from_spec`.  This has the added benefit of understanding more packages through parsing (if they have a complicated package specification) and not requiring the extra step for many of installing in a temp directory to determine package name.

File: tests/test_package_specifier.py [hunk 1]
```diff
@@ -64,3 +67,60 @@ def test_parse_specifier_for_metadata(
...
+    "package_spec_in,pip_args_in,package_spec_expected,pip_args_expected,warning_str",
+    [
+        ('pipx==0.15.0;python_version>="3.6"', [], "pipx==0.15.0", [], None),
+        ("pipx==0.15.0", ["--editable"], "pipx==0.15.0", [], "Ignoring --editable",),
+        (
+            'pipx==0.15.0;python_version>="3.6"',
+            [],
+            "pipx==0.15.0",
+            [],
+            "Ignoring environment markers",
```
Comment (cs01) [inline tests/test_package_specifier.py:82]: Nit: Could the environment markers be captured and printed?

Comment (itsayellow) [reply tests/test_package_specifier.py:82]: Do you mean specifically in the test?  Or do you mean generally for the `pipx` warnings to the user?

Comment (cs01) [reply tests/test_package_specifier.py:82]: Generally for user facing earnings, which would mean the test would also need to be updated. Just a nit, I’m fine merging without the change as well.

Comment (itsayellow) [reply tests/test_package_specifier.py:82]: It sounds like a good idea.  I'll implement it and then merge.

Comment (cs01) [reply tests/test_package_specifier.py]: Looks good. One other thing just came to mind. If I were a user and got that message, I would probably want to know *why* things were being ignored. There are good reasons but they are not always obvious (as your discussion in the related issue revealed). Could reasoning also be added to the warning when arguments are being dropped by pipx?

Comment (itsayellow) [reply tests/test_package_specifier.py]: I can imagine saying for the environment marker case: "Use pipx options instead." as concise but still basically explaining the philosophy.  Does that sound good?

I'm a little at a loss at a concise (warning message length) way to explain ignoring `--editable` for non-local path.  I'm thinking.  Anyone have an idea?

Comment (cs01) [reply tests/test_package_specifier.py]: Sorry for the long delay. The text you ended up with is perfect.

**Pull request merged**